src/ocrd_models/ocrd_page.py

```diff
@@ -174,10 +176,28 @@
     """
 )
 
-# add alias for DOM root
-OcrdPage = PcGtsType
-
-def to_xml(el, skip_declaration=False):
+class OcrdPage():
+    """
+    Proxy object for :py:class:`ocrd_models.PcGtsType` that also offers access
+    to the underlying etree, element-node mapping and reverse mapping, too (cf.
+    :py:func:`ocrd_models.ocrd_page.parseEtree`)
+    """
+    def __init__(
+        self,
+        pcgts : PcGtsType,
+        etree : ET._Element,
+        mapping : dict[str, ET._Element],
+        revmap : dict[ET._Element, str],
+    ):
+        self._pcgts = pcgts
+        self.etree = etree
+        self.mapping = mapping
+        self.revmap = revmap
+
+    def __getattr__(self, name):
+        return getattr(self._pcgts, name)
```

Comment:
Interesting. Will this be enough to yield access to all of PcGtsType's members (functions) as well?

Alternatively, why not subclass it (overriding the constructor) instead?

Reply:
Yes, this will work for all class attributes, including methods and properties.

I prefer composition over inheritance for wrapping an instance when possible because it gives more freedom (you could delegate different calls to different instances e.g.) and you don't depend on the implementation (i.e. we don't need to copy/adapt the constructor). In this simple case, both would work but I also see no reason to use inheritance.

Reply:
The only drawback is that you cannot do `isinstance(OcrdPage, PcGtsType)` anymore.

Reply:
Understood! Thanks for explaining!

Reply:
> The only drawback is that you cannot do `isinstance(OcrdPage, PcGtsType)` anymore.

Ok. So we'll have lots of places to check, esp. in core.

So far, for external code, I can only see `PcGtsType` instantiation in Eynollah und Textract2page as explicit use, and no explicit type checking against that.

Reply:
In core, so far I only see it here: https://github.com/OCR-D/core/blob/18ecf8f06ec4f411fc97c803182fe0be6430a080/src/ocrd_validators/page_validator.py#L239

I guess we'll see the moment we run pytest whether the existing type checks are problematic.

Reply:
I will do another round of refactoring of the OCR-D interface for eynollah, so that should be fixed soon.

> In core, so far I only see it here: https://github.com/OCR-D/core/blob/18ecf8f06ec4f411fc97c803182fe0be6430a080/src/ocrd_validators/page_validator.py#L239

This one I did already catch, tests pass. But it might be cleaner to do

```python
if isinstance(node, OcrdPage):
  node = node.pcgts
```

instead of checking `isinstance(node, (OcrdPage, PcGtsType))` because it makes it clear that while they behave largely the same, an `OcrdPage` is a wrapper for `PcGtsType`.

But that's mostly a syntactic detail, what I still don't know is whether using `parseEtree` is (significantly) more expensive than `parse` and not retaining the mapping. I don't think so, but it would be better if we could measure that.

Reply:
> But it might be cleaner to do

yes!



> what I still don't know is whether using `parseEtree` is (significantly) more expensive than `parse` and not retaining the mapping. I don't think so, but it would be better if we could measure that.

I just did a small side experiment. I took 205 example PAGE-XML files of medium to large complexity (similar to the `Gutachten` test asset, including text annotations on Glyph level) – totalling 112 MB uncompressed on disk – and ran:
```python
>>> from ocrd_models.ocrd_page import parseEtree, parse
>>> import time
>>> from pathlib import Path
>>> p = Path('path/to/gutachten')
>>> def run(method):
    t1 = time.time()
    for f in p.glob('*.xml'):
            _ = method(f, silence=True)
    t2 = time.time()
    return t2-t1
>>> run(parse)
19.130897045135498
>>> run(parseEtree)
22.413432598114014
```

So the effect is there, but it's tolerable. And it will be smaller on smaller files (earlier in the pipeline or fewer content), larger with larger files (say multiple TextEquivs or even labels, or large newspaper pages).